src/derive.rs

```diff
@@ -199,6 +199,63 @@ impl<'a> DerivedModule<'a> {
             new_h
         })
     }
+    // remap a const expression from source context into our derived context
+    pub fn import_const_expression(
+        &mut self,
+        const_expression_handle: &Handle<Expression>,
+    ) -> Handle<Expression> {
+        self.const_expression_map
+            .get(const_expression_handle)
+            .copied()
+            .unwrap_or_else(|| {
+                let old_c = self
+                    .shader
+                    .as_ref()
+                    .unwrap()
+                    .const_expressions
+                    .try_get(*const_expression_handle)
+                    .unwrap();
+
+                let new_const = match old_c {
+                    Expression::Constant(h_c) => {
+                        let c = self
+                            .shader
+                            .as_ref()
+                            .unwrap()
+                            .constants
+                            .try_get(*h_c)
+                            .unwrap();
+
+                        Constant {
+                            name: c.name.clone(),
+                            r#override: c.r#override.clone(),
+                            ty: self.import_type(&c.ty),
+                            init: self.import_const_expression(&c.init),
+                        }
+                    }
+                    // TODO
```

Comment:
const_expressions can be more than just constants, they can be compound expressions as long as all the components are const_expressions as well.

i haven't looked in a lot of detail, but i think rather than making a new function here, the easier thing to do is to use the existing `import_expression` fn passing `self.const_expressions` and `self.shader.const_expressions` in as the new and old arenas (instead of the function-local arenas that are used when import_expression is called in function import).

you will probably need to also make a member for the `already_imported` param, that gets reset when the shader is changed. this might mean we end up with duplicate const_expressions if multiple modules define the same constant, but that should be ok.

Reply:
> use the existing import_expression fn passing self.const_expressions and self.shader.const_expressions in as the new and old arenas

With this approach I'm struggling with the borrow checker :face_with_head_bandage: :face_with_head_bandage: :face_with_head_bandage:  The `import_expression` both borrows `self.const_expressions ` and `self.shader.const_expressions` exclusively. Please give me some time to fix this.

Reply:
I changed them to `RefCell`s but there are many runtime `BorrowMut` errors when running tests, still working at it.

Reply:
i don't think you should change everything to refcell, just the const_expression bits (and the params for import_expression). i should have time at the weekend for a proper look if you get stuck.